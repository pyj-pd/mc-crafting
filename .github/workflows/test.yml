name: Test Code

on:
  push:
    branches-ignore:
      - main
  pull-request:
    branches:
      - '**'

jobs:
  build:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@master
      - name: Install Node.js
        uses: actions/setup-node@v3
        with:
          node-version: 18

      - name: Install pnpm
        uses: pnpm/action-setup@v2
        with:
          version: 7

      - name: Install dependencies and test
        run: |
          pnpm install
          pnpm run test:ci
