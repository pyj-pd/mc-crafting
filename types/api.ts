export interface ApiResponse<Data> {
  data: Data | null
  errorMessage: null | string
  statusCode?: number
}

export const DEFAULT_STATUS_CODE = 200

export type ApiFunction<Response, Argument> = (
  arg: Argument
) => Promise<Response>
