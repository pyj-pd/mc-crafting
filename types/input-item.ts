export type ClickHandler = (id: string, index: number) => void
export type FocusHandler = (index: number) => void
export type BlurHandler = (index: number) => void
