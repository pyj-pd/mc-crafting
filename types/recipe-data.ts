export interface RecipeDataQuery {
  file?: string
  id?: string
}

interface RecipeCommonData {
  type: string
  category?: string
  group?: string
  result: {
    count?: number
    item: string
  }
}

/**
 * Type for recipes with type crafting_shaped
 * @see https://minecraft.fandom.com/wiki/Recipe#crafting_shaped
 */
export interface CraftingShaped extends RecipeCommonData {
  pattern: string[]
  key: {
    [key: string]: {
      item?: string
      tag?: string
    }
  }
}

/**
 * Type for recipes with type crafting_shapeless
 * @see https://minecraft.fandom.com/wiki/Recipe#crafting_shapeless
 */
export interface CraftingShapeless extends RecipeCommonData {
  ingredients: { item?: string; tag?: string }[]
}

/**
 * Default crafting recipe data
 * made by Minecraft itself
 */
export type RawRecipeData = CraftingShaped | CraftingShapeless

/**
 * Crafting recipe data that is converted
 *
 * Pattern is different from Minecraft's one.
 * Patterns contain each key, instead of each row.
 */
export type RecipeData = {
  id: string
  count: number
  key: {
    [key: string]: string[]
  }
  pattern: string[]
  textures: { id: string; texture: string | null }[]
}

/**
 * Item ID and name in pair
 */
export interface ItemNamePair {
  id: string
  file: string
  name: string
  searchName: string
  texture: string
}

/**
 * Data with item ID and its file name
 */
export interface ItemFile {
  id: string
  file: string
}

export type TagValue = string

/**
 * Default tag data
 * made by Minecraft itself
 * @see https://minecraft.fandom.com/wiki/Tag
 */
export interface RawTag {
  replace?: boolean
  values: TagValue[]
}

/**
 * Tag data
 */
export type TagData = TagValue[]
