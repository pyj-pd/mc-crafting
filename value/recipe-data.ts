export const craftingShapelessId = 'minecraft:crafting_shapeless'
export const craftingShapedId = 'minecraft:crafting_shaped'
