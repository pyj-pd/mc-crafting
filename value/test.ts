// All of these data are for testing only

import { RecipeDataQuery } from 'types/recipe-data'

export const ITEM_ID = 'minecraft:white_wool'
export const TAG_ID = 'minecraft:acacia_logs'

export const LANGUAGE_CODE = 'ko'

export const POST_FILE = 'posts/about.md'

export const CRAFTING_RECIPE_ITEMS: Record<
  'shaped' | 'shapeless',
  RecipeDataQuery
> = {
  shaped: { id: ITEM_ID },
  shapeless: { file: 'light_blue_bed_from_white_bed.json' },
}
