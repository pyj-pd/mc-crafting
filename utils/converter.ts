import path from 'path'
import { readFile, readdir } from 'fs/promises'
import {
  ItemFile,
  ItemNamePair,
  RawRecipeData,
  RawTag,
  TagData,
} from 'types/recipe-data'
import { minecraftPrefix } from 'value/minecraft'
import textures from 'minecraft-textures/dist/textures/1.19'
import { getTranslation } from './language'
import { DEFAULT_LANGUAGE_CODE } from 'value/language'
import { craftingShapedId, craftingShapelessId } from 'value/recipe-data'
import hangul from 'hangul-js'
import Inko from 'inko'

/**
 * Disassemble Korean characters and make it one string
 * @param text Text to disassemble
 * @returns Disassembled text in string
 */
export const disassembleHangul = (text: string): string =>
  hangul.d(text).join('')

const inko = new Inko()

/**
 * Convert Korean characters into English
 * @param text Text to convert
 * @returns Converted text
 */
export const koreanToEnglish = (text: string): string => inko.ko2en(text)

/**
 * Convert text to work well with searching
 * @param text Text to convert
 * @returns Converted text
 */
export const textToSearch = (text: string): string =>
  koreanToEnglish(disassembleHangul(text))

const pureIdRegex = new RegExp(`^${minecraftPrefix}`)

/**
 * Get pure ID of item
 * @param id Item ID
 * @returns Item ID without the prefix
 */
const getPureId = (id: string) => id.replace(pureIdRegex, '')

const recipeListDir = path.resolve(process.cwd(), 'data', 'recipes')
const tagListDir = path.resolve(process.cwd(), 'data', 'tags')

/**
 * Get list of recipe items
 * @returns Item IDs in array
 * @deprecated Use `getItemFiles`
 */
export const getRecipeList = async () => {
  const files = await readdir(recipeListDir)

  return files
}

/**
 * Get ID and file name of items
 * @returns Array that contains items' ID and their file name
 */
export const getItemFiles = async (): Promise<ItemFile[]> => {
  const files = await readdir(recipeListDir)

  const data: ItemFile[] = []

  for (const file of files) {
    const fileData = JSON.parse(
      await readFile(path.join(recipeListDir, file), 'utf-8')
    )

    const type = fileData.type
    const isRecipe =
      (type === craftingShapedId || type === craftingShapelessId) &&
      typeof fileData.result.item === 'string'

    if (!isRecipe) continue

    const id = fileData.result.item

    const item: ItemFile = {
      id,
      file,
    }

    data.push(item)
  }

  return data
}

const itemFiles = await getItemFiles()

/**
 * Item list with recipes
 * @param code Language code
 * @param convert Whether to convert names. Used for searching.
 * @returns Array that contains id, name, and texture of each item
 */
export const getItemList = async (
  code = DEFAULT_LANGUAGE_CODE,
  convert = false
): Promise<ItemNamePair[]> => {
  const translation = await getTranslation(code)

  const itemNames: ItemNamePair[] = []

  for (const { id, file } of itemFiles) {
    const itemData = textures.items.find((item) => item.id === id)

    if (!itemData) continue

    const { texture } = itemData

    const name = translation[id]
    const searchName = convert ? textToSearch(name) : name

    const item: ItemNamePair = {
      id,
      file,
      name,
      searchName,
      texture,
    }

    itemNames.push(item)
  }

  return itemNames
}

const itemList = await getItemList()

/**
 * Get crafting recipe
 * @param file Item file name
 * @param id Item ID
 * @returns Crafting recipe of that item
 */
export const getRecipeData = async ({
  file,
  id,
}: {
  file?: string
  id?: string
}) => {
  const item = itemFiles.find((item) =>
    file ? item.file === file : item.id === id
  )

  if (!item) throw new Error('No recipe.')

  const fileDir = path.join(recipeListDir, item.file)

  const data = JSON.parse(await readFile(fileDir, 'utf-8')) as RawRecipeData

  return data
}

/**
 * Get image of the item
 * @param id Item ID
 */
export const getImage = (id: string) =>
  textures.items.find((item) => item.id === id)?.texture ?? null

/**
 * Get list of item and block in a tag
 * @param id Tag ID
 * @returns Array that contains item or block id
 */
export const getTagData = async (id: string): Promise<TagData> => {
  const tagPureId = getPureId(id)
  const tagFileName = `${tagPureId}.json`
  const tagTypes = await readdir(tagListDir)

  let tagFileDir: string | null = null

  for (const dir of tagTypes) {
    // Search through tag files
    const tagList = await readdir(path.join(tagListDir, dir))
    if (tagList.includes(tagFileName)) {
      tagFileDir = path.join(tagListDir, dir, tagFileName)
      break
    }
  }

  if (tagFileDir === null) throw new Error('No such file.')

  const tagFile = JSON.parse(await readFile(tagFileDir, 'utf-8')) as RawTag

  const data: TagData = tagFile.values

  return data
}

/**
 * Check if crafting recipe for the item exists
 * @param id Item ID
 * @returns boolean
 * @deprecated
 */
export const checkRecipeExists = async (id: string) => {
  const exists = itemList.findIndex((item) => item.id === id) !== -1

  return exists
}
