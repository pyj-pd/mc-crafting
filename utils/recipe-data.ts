import {
  CraftingShaped,
  CraftingShapeless,
  RawRecipeData,
  RecipeData,
} from 'types/recipe-data'
import { getImage, getTagData } from './converter'
import { craftingShapelessId } from 'value/recipe-data'

/**
 * Get each texture for items
 * @param ids Array of item IDs
 * @returns Array of item ID and its texture
 */
const getTextures = (ids: string[]) =>
  ids.map((id) => ({
    id,
    texture: getImage(id),
  }))

type ConvertedData = Pick<RecipeData, 'key' | 'pattern' | 'textures'>

const keyNames = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i']
const convertShapelessData = async (
  data: CraftingShapeless
): Promise<ConvertedData> => {
  const {
    ingredients,
    result: { item },
  } = data

  const items: string[] = [item] // Used to add textures

  // Convert ingredients into keys
  const key: ConvertedData['key'] = {}

  for (const [index, item] of ingredients.entries()) {
    const keyName = keyNames[index] // Key name to be used in pattern
    const keyData = item // Key data with item or tag

    const keyItems = []

    if (keyData.item) {
      items.push(keyData.item)
      keyItems.push(keyData.item)
    } else if (keyData.tag) {
      // If it's tag instead of single item,
      // add all items into texture list

      const tagItems = await getTagData(keyData.tag)

      tagItems.forEach((i) => {
        items.push(i)
        keyItems.push(i)
      })
    }

    key[keyName] = keyItems
  }

  const textures = getTextures(items)

  // Make patterns
  const pattern = Object.keys(key)

  return {
    key,
    pattern,
    textures,
  }
}

const convertShapedData = async (
  data: CraftingShaped
): Promise<ConvertedData> => {
  const {
    key: rawKey,
    result: { item },
  } = data

  const items: string[] = [item] // Used to add textures

  const key: ConvertedData['key'] = {}

  for (const keyName in rawKey) {
    const keyData = rawKey[keyName] // Key data with item or tag
    const keyItems = []

    if (keyData.item) {
      items.push(keyData.item)
      keyItems.push(keyData.item)
    } else if (keyData.tag) {
      // If it's tag instead of single item,
      // add all items into texture list

      const tagItems = await getTagData(keyData.tag)

      tagItems.forEach((i) => {
        items.push(i)
        keyItems.push(i)
      })
    }

    key[keyName] = keyItems
  }

  const textures = getTextures(items)

  const pattern = data.pattern.map((row) => row.padEnd(3, ' ').split('')).flat()

  return { key, pattern, textures }
}

/**
 * Converts raw recipe data into key & pattern data
 * @param data Raw crafting recipe data
 * @returns Converted recipe data
 */
export const convertRecipeData = async (
  data: RawRecipeData
): Promise<RecipeData> => {
  const toAdd = {
    count: data.result.count ?? 1,
    id: data.result.item,
  }

  const id = data.type

  if (id === craftingShapelessId) {
    // Shapeless data
    const converted = await convertShapelessData(data as CraftingShapeless)

    return {
      ...toAdd,
      ...converted,
    }
  } else {
    // Shaped data
    const converted = await convertShapedData(data as CraftingShaped)

    return {
      ...toAdd,
      ...converted,
    }
  }
}
