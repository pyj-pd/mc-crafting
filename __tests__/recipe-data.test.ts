import { RecipeDataQuery } from 'types/recipe-data'
import { getRecipeData } from 'utils/converter'
import { convertRecipeData } from 'utils/recipe-data'
import { CRAFTING_RECIPE_ITEMS } from 'value/test'
import { describe, it, expect } from 'vitest'

/**
 * Get recipe data from file or item and convert
 * @returns Converted recipe data
 */
const getRecipeAndConvert = async ({ file, id }: RecipeDataQuery) => {
  const rawRecipeData = await getRecipeData({ file, id })

  const data = await convertRecipeData(rawRecipeData)

  return data
}

describe('Recipe data converter', () => {
  it('converts shaped item', async () => {
    const {
      shaped: { id },
    } = CRAFTING_RECIPE_ITEMS

    const data = await getRecipeAndConvert({ id })

    expect(data.id).toEqual(id)
  })

  it('converts shapeless item', async () => {
    const {
      shapeless: { id, file },
    } = CRAFTING_RECIPE_ITEMS

    const data = await getRecipeAndConvert({ file })

    expect(data.id).toEqual(id)
  })
})
