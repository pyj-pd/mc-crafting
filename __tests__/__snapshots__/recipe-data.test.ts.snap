// Vitest Snapshot v1

exports[`Recipe data converter > converts shaped item 1`] = `
{
  "count": 1,
  "id": "minecraft:white_wool",
  "key": {
    "#": [
      "minecraft:string",
    ],
  },
  "pattern": [
    "#",
    "#",
    " ",
    "#",
    "#",
    " ",
  ],
  "textures": [
    {
      "id": "minecraft:white_wool",
      "texture": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAIIUlEQVR4XmMYaMBIrgMeP3nyn5mJCaz9+8+fDMqKimSZRbKmY6dP/+fi4ARbLC4iDKafvnwJpn/++s1gbWZKkplEK34C9PG3Hz8Z/vz9w8DBxs7AysrCcO/RY7DFEqKiDKwsLAxsbKwMb969Z/j9+zeDqZEhUWYzEYqCMxcu/H/x4sX/P3/+MHz99g3uYxZmZgYeLi4GdSVFsOUsLMwML9+8Zfj89SuDkKAAA0wfIfNxuhIUxy9evQbrZ2dnB9NMQNWvgT6Uk5ZiYGSAaH38/DlYTklOlgGkXlZKkuHnr19gsfuPn4BpKXFxBjUVZax2YQieOHP2Pz8fL1gjL9CHIMYnoK/A9OcvDFLiYgxMwMT39dt3sJr3Hz+CaWGgr38CEyOII8DHBxZ78gKSNkChAxIQExZmkJOVRbGTBRZEoDh+BPQNKC5BFr/7+ImBhQUizcLMAo57djY2BjZWVoZPX74yfPv+nQEULdIS4uBg//bjB8M3oKN+AB0BihoGRkYGGaAcyHxQqEmKiYL1gaLm9+8/DJamJmCHwB2wces2BlZmJgY/Hx+Gm/cfMEgCE9bTl68Y2FlBiYsdaNlfBpAhIDkuYJSICgkycHBwMMBCBuRLkOWg0PkCTCugHAHKJf/+/wfru/f4McO/f/8ZJvb1MfwHisE8DncANycHA8inazdsZOACskUE+IGpnZVBWFAQrBbk4zfv3jFIAX0CCv4PwOj48+EDWI6PlwdM//nzF+xrkI+ZgYn0zfsPDALA6AQ5rL+nl4GDnQ1oBxM4CjEcwMXJCU5WIF/++fuX4fKlSwzvgPGrq6cPVqsoK8PwC5jPQY57Bk2cTIyQ6JQQEYE4hJub4R4w4YFCg5UZ4reG+gaGz9++AkOSleErMNr4uHkY/mELAZAPfwFTLyje+Xl5Gf4CfcPGwspw9fJlBl5uLgYRYJB/+/6D4cWb1wygLCgNTIywbPkTmO8/gUIEmFVBwSshIspQVVXFwA70MSi0QI4BpR2QJ0EuxeoAUPD/ZfkHtIybAZSyQY4AFSwMv/4zfPj0meHC2XMMr969ZTAwNGIAZTlQYgXFKagU/A3M/xLAqAElxAVzZoMTKKiY/g5MmPzA6PkLDFFQlICiApQ2/v3/x4ARBSBLubg4wcHEww1yxCdwaQdSKMTPz/AaGP/swBLwwd07DCBsYGQENgOU6EDlRH1dHTDhMzKAimlmJmag3H8w+zuw9OTkYAfmgC8MIHNBDgCFBoYDQNnp0bPnDKCg/fvvHwMoYb188waY6MQYXr9/D07Jv4FBDBIDhdbhw4cZ/v39x3AESDMDSyhQiIEMBUUB8z8mcHYF5QaQZbAy4zMw+zIB1YKiEMMBoBQuByzFfgHjE5T4fv78xQBSDCpMOIHZ7SMwjkHBKSclBfYpSD0zBzCLArMpE9DHoDTECEyUoGhhBmbnD5+AIQhMQyCLQGkVFGWgNADKKUzMTJhRwAtMnSBDRYSEGDiB+RyEQZUPqGwAZl9wdLz98BGclUCJFaQeFFIgH4IsBoUCKDr+AUMPVnL+Z/gPTg+goAcVPqAQBPn+05cvmA74CBTk4+FmePX2HQM/Dyir/ANnF5AGUPZ5B7QcFBIcQMeBMMhiUOIChRiIBqVskC9BGJQGQA4CqQc5CiQPSgegOgIUUrDoArkCXhAJAwseUDALAWlQ7v71G5gOgIkRpAhkEMghoBAC1XigEu7Zq1cMIINBwQ+K93/AlA4qqkFBD3IgrERkYoQE9x+wPCsDyPIf0DoDxQEgDqgS+gMsp18AEx8o/lmB5T4LMH5BQQfKRpDyQAgcSrDaEKTvBzC9gAofUPCDHAtyyH+gBCjYQebA8v8voNkgT7ID2w0YifDl27fgIlKQl49BHljdgnwHVgwsTP4BKxkWYMIBOQKUt8WEhRhAifQzsMz//uM7gyA/HwPIcnAaAJZyoNQP8jkoDbCwsEHKBWDRzAgMDZBDmcDZFOIEeHIElXoswGr2z7+/DKCgBklzA4tnUHELKjg42DnABRIozkElIgsw9YPKdlCQgkpEcJB/hTRYQGkA1GIC1Yos0BQPshykjhWYM0DpABYCcAfwAItbbmD9/w5YwQgCC573wOqYkQlS1osCcwaoVAT5HBTfoDh88foNOLeADAUF+cfPn8GVGCzvM0LrCVACBmVNkFGgkGIFtpwEoO0NkCPgDgAZAko8spKSYMeBGhivgTkCZNArYFELEgQ5jo2VDWwRqMACOQJUOoLiFlSHgCwS4IU0ZkBOBzXPQPEPCg1Y9IFCDzkbwh3wFNh6AaUDUHYDWQaq+SCl4VuwhW+ApeGfv5AyHJSdQLkAVHqCQgOUxdiBDgPpA1kACjmQ5aCsCiozmKEtKFA0gdTyQnMXSD1Gk2zRkiX/QUUmKEpAcfnj5w9gi+cbuCgGWQyKGnD2YYE0Sg8dPAiuxkE5BSQOshQUHaDyAKQelEZAjgLVM1+ATbtVK1ag2Im1oQgyaNacuf9BcQYr7WD5mwPaQAU1NriARfTGTZvARTaoeQZqhvFycTP8BSZkULkBCnZQGmEDNuUWLlhAXKOUAQ2sXbv2/wdgAgMFGyh6QPEJMgmUFt4CE+yJo0eBOv4zcEMbsCBfg1I8GzCXgLLs0iVLcHoSaxQw4ACz5sz5D7IEFCIg34GLXGDe3gQMAVChAy6OgZUGKJeAWk1z587FazHMGqIUIbtp6oyZ/0GJDFZCnj9zGiwNqgtAPaLFixaRZCbJDoA5ZtXq1f9BmleuXsMAqn5XLl9OtlkMAwkAkRtzGjUCYVQAAAAASUVORK5CYII=",
    },
    {
      "id": "minecraft:string",
      "texture": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAABP0lEQVR4XmMYBdQOgRs3bvwHYWLNZRroKGChlgO+fPnyH2SWVUAQ2EgYn4eHhxGfHcMnBJ48eQL26KU9uxihIfGfmNAdPiEQlpOHEvewENHQ0GAY1GmAkVA8EZuaYeboubihxP2xDevAUrhyw+BNA7DSDJavGYgEsFwAUw7UDw4RmHnANIES6gMeAhhpAL1EQ/cRzCcyMjJgT8JSO4yPK65haQM9TQy+NIBeosHiEj1kcCUJ9BCChQgsJIHmo+SSwRcC6CUazAew3LBqyiSw59FTMyxEgPpRfIie+mFpBaZ+8OUC9BIN5mOYy9HLBfRcgivNwNShh8jgDQFcZTt6iKCXAzA+LC3BfI6rfBn8IYCe39FrO1zlAXqco4fI4M8FDEQC9DIePU3Acg2u3DL00gCugMGVNnD5fNCkAYYRDwBZrKb/DgerKwAAAABJRU5ErkJggg==",
    },
  ],
}
`;

exports[`Recipe data converter > converts shapeless item 1`] = `
{
  "count": 1,
  "id": "minecraft:light_blue_bed",
  "key": {
    "a": [
      "minecraft:white_bed",
    ],
    "b": [
      "minecraft:light_blue_dye",
    ],
  },
  "pattern": [
    "a",
    "b",
  ],
  "textures": [
    {
      "id": "minecraft:light_blue_bed",
      "texture": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAEOUlEQVR4XmMYBSM+BBjJDYGjp0//F+LhQdH+7ssXBmtTU5LMJNsB169f/w+y/datWwysAgIM/FxcDG+ePYM7CCQG43jZ2+O0hywHgHwPMhxkoZqaGoOQkBBqSLx7h8IHORImAFIPCikQHxRazKRGwcyFi/6/ev2Ggen3b7DPP3z9yvDt2zcGIX5+uFGcnJwMyJgLGDqM7OwM3z5/ZhAWFmbgZGNjaG9vZ7h25UojCykOkK3c8r/lBkTH92d34VpbnQQZrly7DubbmZnCxZFDBhRaIlJSYLm2tja4GpIcwCHFx8DBw8nw4dEHBk4lLQaGHz/BBqE4at8usBjIUTBbJAT4waH1ERhSMLFfwBAEsYlOA2JFS8CJTlxXnQEZ/PjyneHHu59wx8Dkfv78wMDOLgDmIoeW48/jDCDLPwOjY9+uXYxEhYBqx6H/MINBvoc7ABgCIItAfH5FOYhDoJLsHOJgloAc0BEc7GD2mw09DL+UeBhAlrN+vAUWI+gAuOXcQPVfGRg4hNiRAoCdgeMrH4T/FUL9/PgSLg8KgQ+3XjKgA45vDxk+//hFnANQNEMd8ePnT7gwyEHgKID5nF+cgQOY4hmgamEKX6ztYWBhYQf7/sXTtwzCfGyEHSDbuBsS9NwIZ3CIczJwMHCiuAuUMBmwAWBBCYqyD9ungWX/fPnI8OLNVQZeLlAoQozGmQhBQQ/zKdxHDMQBmINACfTJvFawz0GW2/E9ZHj+7jvYkEu3X4LtxpkGQD5leAlK3D8J2/oVSQkwtEAWg0WQxE3YbjM8fvWL4dr9NyieZsJmuu780/BUj5rosLsFOU0wQC0FhQIs3kG+f/sJ03KQaRgOgKV6cP4GJjbkBIYrKFAcCQwBkOUPZrYx/PnzkwFkOcj36D6HmYUSHPBEx4Ca3ZATGUrwIqd0qMUgg8GW//wGjL8fDCDLcfkepBYlDbzbNhXsMGZgdvnIDJESD01nwIhTaK6AhQ4sBGDq/gAtB2U5K76bWOMdOSRRQoDbPOA/siTIIQzMaOn07x+wEvGIbAZYfDMglQ/gLPf3LzjoFX9cZcAV9FijgEvP7T8jJxdGVIMdgiMB/P2DyCX/fn2HZzlQ0B+++IxgXYPiPVeBewy7PyhhWPWXkwGro5AthykAJTxQfgdlOQYiAIoDWFjZGazYIfU6NyeizP/6/SfDsZ+aKMb9YfjGgC20GH78YHj84QvBoIcZhuKAP78hwfn52z+Gz9++M3z/9QduqavkPQaQQ0ACMMdhhBZQPzHxjuwTjBBgAPoWZDEzExMDrLgEaQCyGfRUxSGJ9AOkJtNlA5XrkKIE5Kh3H78xHL7+hqR2JtYQAFn489dvjBhEdhBM0lxb+j8vFxMDKPqYWH8xkApQHLDx0B2ym+mg6ONgZiLZAaTrQLMClk5AIcDMyk5/B8CiBRQCf3//ZBhyAAAMfrba7t3LvgAAAABJRU5ErkJggg==",
    },
    {
      "id": "minecraft:white_bed",
      "texture": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAD8UlEQVR4XmMYBSM+BBjJDYEzFy/+FxMURNH+6v17BhN9fZLMJNsBjx49+g+y/fz58wysAgIMYkD89MEDuINAYjCOl709TnvIcgDI9yDDQRYaGhoysLOzo4TEz58/UfggR8IEQOpBIQXig0KLmdQomLlw0f/nL14y/P/xA+zzVx8+MPz984eBn4cHbhQLCwsDMhYVFWVg5uBg+AxUKykpycDNycnQ1NTEcOXy5UYWUhyw7eDB/zD1jx88ZGD48BHMfczwkOEsw0Uw29PRAW4kcsiAQktaQQEs19DQgHAsKQ7QUVQEB/fjFy8YxAwEGEC+R9YPctT2/QcwjJQQ4IeHFkzyGzAEQWyi08DGjRvBvrewsECxABTfoDhFd8xvYHDDEiI4tKC6DuzayQCy/PPnzwz7du1iJCoKbt26hQh6oO9hLgBZCrIIxEdOXGB5YK4A0bISEuAcAmI31tQwMLGxMYAsZ/14C6yMoANglnMAE9EPYLAh530Q+wc0KGE0elZEDxmQrRzfHjJ8/vGLOAcghzfMEZ++f4cLgxwBy1YgQVBC4wOmcphamMLSykq47188fcsgzMdG2AGw/A4yDGYQPz8/AwgjOwwUzAxYACgXgBJsb1cXWBYU9C9un2Pg5QKVG5BYxZkIQUEP8ynMRwxEAlj2AyXQ7OxslHh//g4SepduvwTbjTMNwHyJHNy47IfFP0geFFqwkhBZ/PfLKwwvPv1iuHb/DYqnmbAZ+vLlS3iqR69wsKlHdiTMUlAoIMf7WyyWg8zCcAAs1YN8ATIYOYHhCgFkR4JCAGR5QVERw79fv8BZDuR7dJ/DzEIJDliiA0kiG4pcpCIHL3JKh1kM0guy/MunTwxfgZUSyHJcvgepRUkDXa2tYIcxAWs3bmC+B3Hqa2sZ0OMUlitgoQNzLEwdyHJQgcP66jLWeEcOSZQQCAsL+48siewQmPhXaMEDciwsvpHLB1CWgxW1oCyHK+ixRoGXn99/HqjP0R3CgAP8Q6r7vwMLKFhRCwr6wxefEaxrUKKA9cM1hi8CWhhW8WCxHBQ6/9AaHiBloIQHKudBWY6BCIDiABZWdoavz8+DtXFzIlo5777/ZGCTNEQ1DhgV2EILlPBevPrCQCjoYYahOODPb0hT6vO3fwyfv31n+P7rD9xSRc5rDF+//0RxHHpo/QCGCDHxjuwTjBBgAFoCspiZiYkBVlyCNADZDHqq4pBE+gFSk3G+PQss15ngjvr88RvRPscbAiDJn79+Y8QgsoNgkuba0v95uZgYQNHHxPqLgVSAEgIbD90hu5kOij4OZiaSHUC6DjQrYOkEFALMrOz0dwAsWkAh8Pf3T4YhBwCt1uRJfp703AAAAABJRU5ErkJggg==",
    },
    {
      "id": "minecraft:light_blue_dye",
      "texture": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAABJUlEQVR4XmMYBaMhMBoCIz4EGIkNAZvUGf+xqT0yOwNsBro8TJyQ+UwDHQUEQwDms+q8FLBbrz35CqYPnbwNpt8+PQOmA4MiwfTDR8/B9Ibls8D0owN9eO0Y8BBgwRUFMJ/DfDZj7QWwUnlJHhQtMHktGW6ouCSYNnSIgqkDpx1cITH40gCuOIfF7cPnX8A+szNXBdOfP38G0+/ef0EJGZg6mOD5A8vATPSQGLxpAJbaCfncXFMC7DP03AFLK+ghgZ7mBm8IwPI5zCe44hxX7iDkc1hIDN6SUM6hCJx/YfkZFhKeVioocY6eRtDjGFfqHzQhgLMkRMqv4JA4D3Xyw+dRDPgAzMcwNYO+LiC6PQDzESxt4AoFQj4edOUAwygYDYERHwIA8TR/tTzXKj0AAAAASUVORK5CYII=",
    },
  ],
}
`;
